nntrainer/opencl/opencl_program.cpp

```diff
@@ -41,6 +43,105 @@ bool Program::BuildProgram(cl_device_id device_id,
     return false;
   }
 
+  // saving kernel binary
+  if (!binaryCreated)
+    return GetProgramInfo(device_id);
+
+  return true;
+}
+
+/**
+ * @brief Utility to get program info and save kernel binaries
+ *
+ * @param device_id OpenCL device id
+ * @return true if successful or false otherwise
+ */
+bool Program::GetProgramInfo(cl_device_id device_id) {
+  // since only GPU is being used
+  unsigned int num_devices = 1;
+
+  cl_int error_code = CL_SUCCESS;
+  size_t *binaries_size = NULL;
+  unsigned char **binaries_ptr = NULL;
+  char *kernel_names;
+
+  // Read the binary size
+  size_t binaries_size_alloc_size = sizeof(size_t) * num_devices;
+  binaries_size = (size_t *)malloc(binaries_size_alloc_size);
+  error_code = clGetProgramInfo(program_, CL_PROGRAM_BINARY_SIZES,
+                                binaries_size_alloc_size, binaries_size, NULL);
+
+  if (error_code != CL_SUCCESS) {
+    ml_loge("Failed to get program binary size. OpenCL error code: %d. %s",
+            error_code,
+            (GetProgramBuildInfo(device_id, CL_PROGRAM_BUILD_LOG)).c_str());
+    return false;
+  }
+
+  // Read the kernel name
+  size_t *kernel_names_size;
+  error_code = clGetProgramInfo(program_, CL_PROGRAM_KERNEL_NAMES, 0, NULL,
+                                kernel_names_size);
+
+  if (error_code != CL_SUCCESS) {
+    ml_loge("Failed to get program kernel name size. OpenCL error code: %d. %s",
+            error_code,
+            (GetProgramBuildInfo(device_id, CL_PROGRAM_BUILD_LOG)).c_str());
+    return false;
+  }
+
+  // since only one kernel is being saved via this function at a time
+  // kernel_names_size -> array with 1 element
+  kernel_names = (char *)malloc(kernel_names_size[0]);
+  error_code = clGetProgramInfo(program_, CL_PROGRAM_KERNEL_NAMES,
+                                kernel_names_size[0], kernel_names, NULL);
```

Comment:
You will be accessing either a null pointer or a garbage value.

Reply:
And ```kernel_names_size[0]``` gives the impression that ```kernel_names_size``` is an array with multiple elements. When it's a pointer to a single value, please write ```*kernel_names_size```.